src/interface/container.cpp

```diff
@@ -147,6 +143,221 @@ Container<T> Container<T>::SparseSlice(int id) {
   return c;
 }
 
+/// Queries related to variable packs
+/// TODO(JMM): Make sure this is thread-safe
+/// TODO(JMM): Should the vector of names be sorted to enforce uniqueness?
+template <typename T>
+VariableFluxPack<T> Container<T>::PackVariablesAndFluxes_(
+    const std::vector<std::string> &var_names, const std::vector<std::string> &flx_names,
+    const vpack_types::VarList<T> &vars, const vpack_types::VarList<T> &fvars,
+    PackIndexMap &vmap) {
+  auto key = std::make_pair(var_names, flx_names);
+  auto kvpair = varFluxPackMap_.find(key);
+  if (kvpair == varFluxPackMap_.end()) {
+    auto pack = MakeFluxPack(vars, fvars, &vmap);
+    FluxPackIndxPair<T> value;
+    value.pack = pack;
+    value.map = vmap;
+    varFluxPackMap_[key] = value;
+    // varFluxPackMap_[key] = std::make_pair(pack,vmap);
+    return pack;
+  }
+  vmap = (kvpair->second).map;
+  return (kvpair->second).pack;
+  // vmap = std::get<1>(kvpair->second);
+  // return std::get<0>(kvpair->second);
+}
+
+template <typename T>
+VariableFluxPack<T>
+Container<T>::PackVariablesAndFluxes(const std::vector<std::string> &var_names,
+                                     const std::vector<std::string> &flx_names,
+                                     PackIndexMap &vmap) {
+  static std::vector<std::string> all_names;
+  static std::vector<std::string> all_flux_names;
+  vpack_types::VarList<T> vars = MakeList_(var_names, all_names);
+  vpack_types::VarList<T> fvars = MakeList_(flx_names, all_flux_names);
+  return PackVariablesAndFluxes_(all_names, all_flux_names, vars, fvars, vmap);
```

Comment:
I will add some documentation. That's right. The variables for the pack are used for the key of the cache, not the variables in the container. 

The issue is that the names of the variable is actually not sufficient, because a pack may contain only a subset of sparse indices for a sparse variable. So `all_names` contains the names/labels of all the variables included in the pack, but the sparse variable is expanded to `sparse_idx` where `sparse` is the name of the variable and `idx` is the index of the specific field in the sparse variable being extracted. Making it static is a performance enhancement to do with allocation. I wanted the vector size to persist between function calls. I'm not sure it makes a difference and it is a little confusing, so I'll remove the static.

Reply:
I renamed `all_names` to `expanded_names`, which is a bit more clear, since they expand the sparse variable names to include indices.